plugins/emoji/plugin.js

```diff
@@ -79,26 +152,466 @@
-								return -1;
-							} else {
-								return 1;
-							}
 						} );
 					callback( data );
 				}
 			} );
+
+			editor.addCommand( 'insertEmoji', {
+				exec: function( editor, data ) {
+					editor.insertHtml( data.emojiText );
+				}
+			} );
+
+			if ( !editor.plugins.toolbar ) {
+				return;
+			}
+
+			editor.ui.addToolbarGroup( 'emoji', 'insert' );
+			// Name is responsible for icon name also.
+			editor.ui.add( 'emojiPanel', CKEDITOR.UI_PANELBUTTON, {
+				label: 'emoji',
+				title: 'Emoji List',
+				modes: { wysiwyg: 1 },
+				editorFocus: 0,
+				toolbar: 'insert',
+				panel: {
+					css: [ CKEDITOR.skin.getPath( 'editor' ), this.path + 'skins/default.css' ],
+					attributes: { role: 'listbox', 'aria-label': 'Emoji List' },
+					markFirst: false
+				},
+
+				onBlock: function( panel, block ) {
+					var keys = block.keys,
+						rtl = editor.lang.dir === 'rtl';
+
+					keys[ rtl ? 37 : 39 ] = 'next'; // ARROW-RIGHT
+					keys[ 40 ] = 'next'; // ARROW-DOWN
+					keys[ 9 ] = 'next'; // TAB
+					keys[ rtl ? 39 : 37 ] = 'prev'; // ARROW-LEFT
+					keys[ 38 ] = 'prev'; // ARROW-UP
+					keys[ CKEDITOR.SHIFT + 9 ] = 'prev'; // SHIFT + TAB
+					keys[ 32 ] = 'click'; // SPACE
+
+					blockElement = block.element;
+					block.element.getAscendant( 'html' ).addClass( 'cke_emoji' );
+					block.element.getDocument().appendStyleSheet( CKEDITOR.getUrl( CKEDITOR.basePath + 'contents.css' ) );
+					block.element.addClass( 'cke_emoji-panel_block' );
+					block.element.setHtml( createEmojiBlock() );
+					panel.element.addClass( 'cke_emoji_panel' );
+
+					blockObject = block;
+
+					registerListeners( listeners );
+				},
+
+				onOpen: openReset()
+			} );
+
+			function createEmojiBlock() {
+				var output = [];
+
+				output.push( createGroupsNavigation() );
+				output.push( createSeparator() );
+				output.push( createSearchSection() );
+				output.push( createEmojiListBlock() );
+				output.push( createSeparator() );
+				output.push( createStatusBar() );
+
+				return '<div class="cke_emoji_inner_panel">' + output.join( '' ) + '</div>';
+			}
+
+			function createSeparator() {
+				return '<hr>';
+			}
+
+			function createGroupsNavigation() {
+				var itemTemplate,
+					items,
+					svgUrl,
+					imgUrl;
+
+				if ( !CKEDITOR.env.ie ) {
+					svgUrl = CKEDITOR.getUrl( that.path + 'assets/icons-all.svg' );
+
+					itemTemplate = new CKEDITOR.template(
+						'<li class="cke_emoji-navigation_item" data-cke-emoji-group="{group}"><a href="{href}" draggable="false" _cke_focus="1"><svg viewBox="0 0 34 34"> <use href="' +
+						svgUrl +
+						'{href}"></use></svg></a></li>'
+					);
+
+					items = CKEDITOR.tools.array.reduce( GROUPS, function( acc, item ) {
+						return acc + itemTemplate.output( {
+							group: escapeString( item.name ),
+							href: escapeString( '#' + item.name.toLowerCase() )
+						} );
+					}, '' );
+				} else {
+					imgUrl = CKEDITOR.getUrl( that.path + 'assets/icons-all.png' );
```

Comment:
Sadly we don't use separators in file name. It's a pain, but we need to keep it consistent with the rest of the editor.